src/azuqua.js

```diff
@@ -0,0 +1,847 @@
+'use strict';
+
+// Standard Lib modules
+const url = require('url');
+const crypto = require('crypto');
+const querystring = require('querystring');
+const path = require('path');
+const fs = require('fs');
+
+// External modules
+const fetch = require('node-fetch');
+const Promise = require('bluebird');
+const _ = require('lodash');
+
+// Use bluebird for promises
+fetch.Promise = Promise;
+
+// Local Requires
+// Explination: Conditionally load routes try local but if run from src, load from another path
+const routes = (() => {
+  try {
+    return require('./static/routes')
+  } catch (e) {
+    return require('../static/routes')
+  }
+})();
+
+/**
+ * `AzuquaCallback`
+ * @callback azuquaCallback
+ * @param {object} Error - An error object containing information about the error
+ * @param {object} Response - A response object containing relevant response information
+ */
+
+// Checks if the response is within error range and if it is - throws an error
+// Error could possible not be JSON so do what we do with toJSON() and first get text then parse
+// If it's not JSON reject with the text and let the error handler handle it
+function checkResponseError(response) {
+  if (response.status < 200 || response.status >= 400) {
+    return response.text()
+      .then((text) => {
+        try {
+          let responseJson = JSON.parse(text);
+          return Promise.reject(responseJson);
+        } catch (e) {
+          let errProxy = { error : text };
+          errProxy.name = 'Azuqua server response error (statusCode < 200 || statusCode > 400)';
+          return Promise.reject(errProxy);
+        }
+      });
+  } else {
+    return response;
+  }
+};
+
+// Promise 'middleware' takes the fetch response and converts it to a JS object
+function toJSON(response) {
+  return response.text()
+    .then(text => {
+      try {
+        let responseJson = JSON.parse(text);
+        return Promise.resolve(responseJson);
+      } catch (e) {
+        return Promise.reject({ 
+          type : 'ResponseParsingError',
+          message : `Error trying to parse the following server response as JSON: ${text}`
+        });
+      }
+    }) 
+}
```

Comment:
Added b/c I wanted a way to add some additional information if the parsing fails. Suppose I could catch that error down the line though.